src/tfs/tlog.c

```diff
@@ -471,14 +479,14 @@ closure_function(2, 1, void, log_switch_complete,
 void log_flush(log tl, status_handler completion)
 {
     tlog_debug("%s: log %p, completion %p, dirty %d\n", __func__, tl, completion, tl->dirty);
-    if (!tl->dirty) {
+    if (!tl->dirty && !tl->compacting) {
         if (completion)
             apply(completion, STATUS_OK);
         return;
     }
     if (completion)
         vector_push(tl->flush_completions, completion);
-    if (tl->flushing)
+    if (tl->flushing || tl->compacting)
         return;
```

Comment:
This is what may cause the log's tuple_staging buffer to fill up during a log compaction and consequently log_write_eav() to fail, leading to a NOSPACE filesystem error when doing "heavy" file write operations (#1473).
It seems to me that an ongoing log compaction shouldn't prevent nor delay a log flush.

Reply:
I agree that the check for tl->compacting here shouldn't be necessary, but I don't think that is the culprit, as compacting is set on the original log being compacted, whereas the log that is overflowing is fs->temp_log, which is set after log_write is called from filesystem_log_rebuild. But fs->temp_log->flushing is set during the long duration of the backlogged log write (see https://github.com/nanovms/nanos/issues/1473#issuecomment-840823845), and fs->temp_log->tuple_staging extends past the log extension size, causing log_write_eav to fail.

While the short-term issue is the poor performance of virtio-storage (#1483), the hard failure on a log write can be fixed by removing the tuple_staging size limit in log_write_eav and log_write. These checks were probably left over from an earlier version of tlog that did not allow tuple encodings to span multiple log extensions. Now that they can, I think these hard limits are probably unnecessary.

Removing these checks allows the fs_full test to get further with STORAGE=virtio-blk, however the system runs out of memory as neither the log data nor queued data is making it out to disk due to the storage issue.

Reply:
The checks in log_write_eav and log_write have been added because otherwise there is no way for the kernel to report out-of-space errors to userspace for filesystem operations that don't involve allocations of new extents (e.g. creating directories or empty files). While those checks can certainly be improved, I still think the staging tuple cannot be allowed to grow indefinitely in these functions.

Reply:
But those checks don't actually have anything to do with an out-of-storage condition and could cause a failure on a transient condition, such as a backlog of storage requests. It makes sense that log_write_eav and log_write should fail on out of storage conditions, though. Right now it seems that a storage allocation failure would result in an error status being passed to a log_flush completion, which could in turn be passed to a filesystem_flush completion or, in some cases like on flush timer or in log_set_dirty, just be silently ignored.

Perhaps such an error should be recorded in the log structure and checked by future calls to log_write / log_write_eav. And some hard limit for tuple_staging can be enforced, too, so that allocations aren't runaway if a storage device becomes critically unresponsive (such as in the current situation with virtio-blk).

Reply:
Recording a storage allocation failure in the log structure and checking it in future calls to log_write could be an option, with the caveat that such error would only be reported to userspace at a later syscall, whereas the syscall that actually should have returned the error return success instead. So things such as mkdir() or rename() could return success but not be actually recorded to disk.